### ISSUE TITLE
Performance issue depending on response body size

### ISSUE THREAD

**Issue by davekaro** [socketry/falcon#137]
After learning about the new FiberScheduler interface in Ruby 3, I wanted to learn more about real world usage for web applications. This quickly led me to Falcon and the many great socketry gems. Thanks for these!

I thought it would be interesting to take some of the puma benchmarks found https://github.com/puma/puma/tree/master/benchmarks/wrk and compare with Falcon on Ruby 3. 

I found that when running `wrk -c 1 -t 1 -d 10 --latency https://localhost:9292` against puma with a rackup file like:

```
require "securerandom"

response = SecureRandom.hex(8_117)

run lambda { |env|
  [200, {"Content-Type" => "text/plain"}, [response.dup]]
}
```

results in average latency of 436.11us and 2289.53 requests per second. But the same with falcon results in average latency of 50.80ms and 19.60 requests per second.

I created this gist https://gist.github.com/davekaro/d5327431eabac1d4e17a079342dc1b74 with all my code and the results. In an effort to isolate variables (concurrency issues), I ran `wrk` with 1 connection and 1 thread. I set puma to 1 worker and 1 thread and I set Falcon to 1 instance. Two things to note:

1. If the response body size is <= `SecureRandom.hex(8_116)`, Falcon performs equal to Puma.
2. If the request is not using SSL, Falcon performs equal to Puma.

The fact that going from returning a response of `SecureRandom.hex(8_117)` to `SecureRandom.hex(100_000)` perform about the same (~50ms latency), but a response of `SecureRandom.hex(8_116)` is so much faster (~0.3ms) makes me think I either misunderstand something, or we hit some kind of limit around 8kb that results in a different code path.

**Comment by ioquatix** [comment]
> If the request is not using SSL, Falcon performs equal to Puma.

So you compared Puma SSL and Falcon SSL and found a performance difference?

**Comment by davekaro** [comment]
> So you compared Puma SSL and Falcon SSL and found a performance difference?

Yes, I started Puma like:

```
puma -w 1 -t 1 -b "ssl://localhost:9292?key=puma_keypair.pem&cert=cert_puma.pem&verify_mode=none" config.ru
```

I found with Puma, performance degraded from ~0.2ms with a tiny response body to ~2ms with ~100kb response body. Falcon degraded from ~0.3ms to ~50ms in the same test.

**Comment by ioquatix** [comment]
I have not compared the performance of SSL between the two servers before, so this is an interesting result.

50ms vs 2ms is very big difference.

On my laptop, which is not particularly great, I am not able to reproduce your results. At least, in my simple case.

Here is starting falcon:

```
> RESPONSE_SIZE=8117 falcon --quiet serve --forked --count 1
  0.0s     info: Falcon::Command::Serve [oid=0x9d8] [pid=37242] [2021-01-01 11:22:57 +1300]
               | Falcon v0.37.1 taking flight! Using Async::Container::Forked {:count=>1}.
               | - Binding to: #<Falcon::Endpoint https://localhost:9292/ {}>
               | - To terminate: Ctrl-C or kill 37242
               | - To reload configuration: kill -HUP 37242
/Users/samuel/Documents/socketry/falcon/lib/falcon/version.rb:24: warning: already initialized constant Falcon::VERSION
/Users/samuel/.gem/ruby/3.0.0/gems/falcon-0.37.1/lib/falcon/version.rb:24: warning: previous definition of VERSION was here
 0.21s     info: Falcon::Controller::Serve [oid=0xa50] [pid=37242] [2021-01-01 11:22:58 +1300]
               | Starting Falcon Server on https://localhost:9292/
response.bytesize = 16234
 0.22s     info: Async::Container::Process::Instance [oid=0xa64] [pid=37243] [2021-01-01 11:22:58 +1300]
               | - Per-process status: kill -USR1 37243
```

I printed out the `response.bytesize` for informational purposes.

Then running wrk:

```
> wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   205.19us  202.14us  10.62ms   98.89%
    Req/Sec     4.99k   168.42     5.17k    92.08%
  Latency Distribution
     50%  183.00us
     75%  205.00us
     90%  260.00us
     99%  428.00us
  50129 requests in 10.10s, 783.33MB read
Requests/sec:   4963.27
Transfer/sec:     77.56MB
```

I compare it with Puma:

```
puma -w1 -t1 -b "ssl://localhost:9292?key=$HOME/.localhost/localhost.key&cert=$HOME/.localhost/localhost.crt&verify_mode=none"
[37309] Puma starting in cluster mode...
[37309] * Puma version: 5.1.1 (ruby 3.0.0-p0) ("At Your Service")
[37309] *  Min threads: 1
[37309] *  Max threads: 1
[37309] *  Environment: development
[37309] *   Master PID: 37309
[37309] *      Workers: 1
[37309] *     Restarts: (✔) hot (✔) phased
[37309] * Listening on ssl://localhost:9292?key=/Users/samuel/.localhost/localhost.key&cert=/Users/samuel/.localhost/localhost.crt&verify_mode=none
[37309] Use Ctrl-C to stop
response.bytesize = 16384
[37309] - Worker 0 (PID: 37310) booted, phase: 0
```

```
> wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   376.63us  109.07us   4.90ms   91.88%
    Req/Sec     2.66k    62.36     2.74k    92.08%
  Latency Distribution
     50%  346.00us
     75%  378.00us
     90%  446.00us
     99%  699.00us
  26735 requests in 10.10s, 419.47MB read
Requests/sec:   2647.06
Transfer/sec:     41.53MB
```

Here is my `config.ru`:

```
> cat config.ru 

require "securerandom"

response = SecureRandom.hex(ENV.fetch('RESPONSE_SIZE', 1024*8).to_i)

puts "response.bytesize = #{response.bytesize}"

run lambda { |env|
	[200, {"Content-Type" => "text/plain"}, [response.dup]]
}
```

**Comment by ioquatix** [comment]
By the way, I used same key pair for both tests (generated by localhost gem). You can do the same.

**Comment by ioquatix** [comment]
Oh, I didn't compare exactly the same response body size for Puma, let me try it...

```
> RESPONSE_SIZE=8117 puma -w1 -t1 -b "ssl://localhost:9292?key=$HOME/.localhost/localhost.key&cert=$HOME/.localhost/localhost.crt&verify_mode=none"
[37369] Puma starting in cluster mode...
[37369] * Puma version: 5.1.1 (ruby 3.0.0-p0) ("At Your Service")
[37369] *  Min threads: 1
[37369] *  Max threads: 1
[37369] *  Environment: development
[37369] *   Master PID: 37369
[37369] *      Workers: 1
[37369] *     Restarts: (✔) hot (✔) phased
[37369] * Listening on ssl://localhost:9292?key=/Users/samuel/.localhost/localhost.key&cert=/Users/samuel/.localhost/localhost.crt&verify_mode=none
[37369] Use Ctrl-C to stop
response.bytesize = 16234
[37369] - Worker 0 (PID: 37370) booted, phase: 0
```

Results (it was slower):

```
> wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   395.09us  127.97us   5.09ms   90.09%
    Req/Sec     2.54k   157.24     2.72k    86.14%
  Latency Distribution
     50%  358.00us
     75%  401.00us
     90%  522.00us
     99%  775.00us
  25545 requests in 10.10s, 397.14MB read
Requests/sec:   2529.32
Transfer/sec:     39.32MB
```

**Comment by ioquatix** [comment]
```
> openssl version
LibreSSL 2.8.3
```

Is it the same as you?

**Comment by davekaro** [comment]
This is fascinating. I use your example, start up falcon `RESPONSE_SIZE=8116 falcon --quiet serve --forked --count 1` and run wrk:

```
> wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   322.41us    1.57ms  42.47ms   99.60%
    Req/Sec     4.24k   325.75     5.01k    88.12%
  Latency Distribution
     50%  212.00us
     75%  260.00us
     90%  320.00us
     99%  542.00us
  42578 requests in 10.10s, 665.24MB read
Requests/sec:   4215.91
Transfer/sec:     65.87MB
```

Then change it to `RESPONSE_SIZE=8117 falcon --quiet serve --forked --count 1` and wrk again:

```
wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    51.06ms    3.09ms  60.07ms   89.34%
    Req/Sec    19.50      2.18    20.00     95.05%
  Latency Distribution
     50%   50.00ms
     75%   50.06ms
     90%   59.92ms
     99%   60.05ms
  197 requests in 10.10s, 3.09MB read
Requests/sec:     19.50
Transfer/sec:    313.68KB
```

Doing the same with puma, I get:

```
wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   421.85us  135.72us   2.01ms   82.84%
    Req/Sec     2.39k   142.95     2.68k    71.29%
  Latency Distribution
     50%  382.00us
     75%  472.00us
     90%  588.00us
     99%    0.92ms
  23961 requests in 10.10s, 372.52MB read
Requests/sec:   2372.45
Transfer/sec:     36.88MB
```
and
```
wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   468.97us  150.05us   2.22ms   83.17%
    Req/Sec     2.14k   194.02     2.55k    68.00%
  Latency Distribution
     50%  427.00us
     75%  524.00us
     90%  649.00us
     99%    1.02ms
  21343 requests in 10.00s, 331.77MB read
Requests/sec:   2134.26
Transfer/sec:     33.18MB
```

My openssl is

```
openssl version
OpenSSL 1.1.1i  8 Dec 2020
```

I'm running Ubuntu 20.04 on WSL2. I will try a regular Ubuntu box soon.

**Comment by ioquatix** [comment]
Ohhhhhh... WSL2... I wouldn't be surprised if that was causing problems.

**Comment by davekaro** [comment]
Ok I tried on another machine I had on Digital Ocean:

```
> lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.4 LTS
Release:        18.04
Codename:       bionic

> openssl version
OpenSSL 1.1.1  11 Sep 2018
```

With `RESPONSE_SIZE=8116 falcon --quiet serve --forked --count 1` I get:

```
wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
connection 0: 37905 requests completed
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   374.02us    1.85ms  47.33ms   99.52%
    Req/Sec     3.77k   450.19     4.00k    96.04%
  Latency Distribution
     50%  242.00us
     75%  264.00us
     90%  341.00us
     99%  594.00us
  37905 requests in 10.10s, 592.23MB read
Requests/sec:   3752.72
Transfer/sec:     58.63MB
```

With `RESPONSE_SIZE=8117 falcon --quiet serve --forked --count 1` I get:

```
wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
connection 0: 228 requests completed
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    44.02ms  376.17us  49.54ms   98.68%
    Req/Sec    22.57      4.62    30.00     72.28%
  Latency Distribution
     50%   44.00ms
     75%   44.03ms
     90%   44.05ms
     99%   44.24ms
  228 requests in 10.10s, 3.56MB read
Requests/sec:     22.57
Transfer/sec:    361.10KB
```

Puma is slower but shows no slow down between the two different response sizes.

**Comment by ioquatix** [comment]
I could reproduce the problem locally on Linux but not macOS. Odd. I'll take a look at what is going on.

**Comment by ioquatix** [comment]
Okay, it turns out I was explicitly disabled flow control/sync. I've fixed it.

```
> curl --insecure https://localhost:9292 | wc -c
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 16400  100 16400    0     0  1601k      0 --:--:-- --:--:-- --:--:-- 1601k
16400

> wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   133.14us   39.24us   1.62ms   92.13%
    Req/Sec     7.54k   270.51     8.20k    80.20%
  Latency Distribution
     50%  124.00us
     75%  140.00us
     90%  165.00us
     99%  280.00us
  75731 requests in 10.10s, 1.17GB read
Requests/sec:   7498.30
Transfer/sec:    118.36MB
```

Compared with puma:

```
> wrk -c 1 -t 1 -d 10 --latency https://localhost:9292
Running 10s test @ https://localhost:9292
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   234.83us   52.64us   1.79ms   86.14%
    Req/Sec     4.25k   388.88     4.75k    79.21%
  Latency Distribution
     50%  216.00us
     75%  251.00us
     90%  308.00us
     99%  415.00us
  42759 requests in 10.10s, 671.53MB read
Requests/sec:   4233.78
Transfer/sec:     66.49MB
```

This was released in `async-http v0.54.1`. Thanks for bringing this issue to my attention.
